File: problems/2594_minimum_time_to_repair_cars/Solution.java
class Solution {
    public long repairCars(int[] ranks, int cars) {
        long lo = 1, hi = (long) ranks[0] * cars * cars;
        for (int r : ranks) hi = Math.min(hi, (long) r * cars * cars);
        while (lo < hi) {
            long mid = (lo+hi)/2, total = 0;
            for (int r : ranks) total += (long) Math.sqrt(mid / (double) r);
            if (total >= cars) hi = mid; else lo = mid+1;
        }
        return lo;
    }
}

File: problems/2594_minimum_time_to_repair_cars/solution.py
from typing import List


class Solution:
    def repairCars(self, ranks: List[int], cars: int) -> int:
        lo, hi = 1, min(ranks) * cars * cars
        while lo < hi:
            mid = (lo + hi) // 2
            total = sum(int((mid // r) ** 0.5) for r in ranks)
            if total >= cars: hi = mid
            else: lo = mid + 1
        return lo
